File: src/main/java/com/freya02/botcommands/api/pagination/menu/BasicMenu.java
package com.freya02.botcommands.api.pagination.menu;

import com.freya02.botcommands.api.new_components.InteractionConstraints;
import com.freya02.botcommands.api.pagination.PaginatorSupplier;
import com.freya02.botcommands.api.pagination.TimeoutInfo;
import com.freya02.botcommands.api.pagination.paginator.BasicPaginator;
import com.freya02.botcommands.api.pagination.transformer.EntryTransformer;
import com.freya02.botcommands.api.utils.ButtonContent;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.MessageEmbed;
import net.dv8tion.jda.internal.utils.Checks;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @param <E> Type of the menu elements
 * @param <T> Type of the implementor
 */
public abstract class BasicMenu<E, T extends BasicMenu<E, T>> extends BasicPaginator<T> {
	protected final Map<Integer, MenuPage<E>> pages;

	protected BasicMenu(InteractionConstraints constraints,
	                    TimeoutInfo<T> timeout,
	                    boolean hasDeleteButton,
	                    ButtonContent firstContent,
	                    ButtonContent previousContent,
	                    ButtonContent nextContent,
	                    ButtonContent lastContent,
	                    ButtonContent deleteContent,
	                    @NotNull Map<Integer, MenuPage<E>> pages,
	                    @Nullable PaginatorSupplier<T> supplier) {
		super(constraints, timeout, pages.size(), supplier, hasDeleteButton, firstContent, previousContent, nextContent, lastContent, deleteContent);

		this.pages = pages;
	}

	@SuppressWarnings("unchecked")
	@Override
	@NotNull
	protected MessageEmbed getEmbed() {
		final EmbedBuilder builder;

		if (supplier != null) {
			builder = new EmbedBuilder(supplier.get((T) this, messageBuilder, components, page));
		} else {
			builder = new EmbedBuilder();
		}

		final MenuPage<E> menuPage = pages.get(page);

		builder.appendDescription(menuPage.content());

		return builder.build();
	}

	@NotNull
	protected static <E> Map<Integer, MenuPage<E>> makePages(@NotNull List<E> entries,
	                                                         @NotNull EntryTransformer<? super E> transformer,
	                                                         @NotNull RowPrefixSupplier rowPrefixSupplier,
	                                                         int maxEntriesPerPage) {
		final Map<Integer, MenuPage<E>> pages = new HashMap<>();

		int page = 0;
		int oldEntry = 0;
		StringBuilder builder = new StringBuilder();

		for (int i = 0, entriesSize = entries.size(); i < entriesSize; i++) {
			E entry = entries.get(i);

			final String s = transformer.toString(entry);
			Checks.notLonger(s, MessageEmbed.TEXT_MAX_LENGTH - 8, "Entry #" + i + " string");

			if (i - oldEntry >= maxEntriesPerPage || builder.length() + s.length() > MessageEmbed.TEXT_MAX_LENGTH - 8) {
				pages.put(page, new MenuPage<>(builder.toString(), entries.subList(oldEntry, i)));

				page++;
				oldEntry = i;

				builder.setLength(0);
			}

			builder.append(rowPrefixSupplier.apply(i - oldEntry + 1, maxEntriesPerPage)).append(s).append('\n');
		}

		pages.put(page, new MenuPage<>(builder.toString(), entries.subList(oldEntry, entries.size())));

		return pages;
	}
}

File: src/main/java/io/github/freya022/botcommands/api/pagination/menu/BasicMenu.kt
package io.github.freya022.botcommands.api.pagination.menu

import io.github.freya022.botcommands.api.components.Components
import io.github.freya022.botcommands.api.pagination.paginator.BasicPaginator
import io.github.freya022.botcommands.api.pagination.transformer.EntryTransformer
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.internal.utils.Checks

/**
 * @param E Type of the menu elements
 * @param T Type of the implementor
 */
abstract class BasicMenu<E, T : BasicMenu<E, T>> protected constructor(
    componentsService: Components,
    builder: BasicMenuBuilder<*, *, T>,
    protected val pages: Map<Int, MenuPage<E>>
) : BasicPaginator<T>(
    componentsService,
    builder
) {
    override var maxPages: Int = pages.size

    override fun getEmbed(): MessageEmbed {
        val builder = when {
            supplier != null -> EmbedBuilder(supplier.get(this as T, messageBuilder, components, page))
            else -> EmbedBuilder()
        }

        val menuPage = pages[page]!!

        builder.appendDescription(menuPage.content)

        return builder.build()
    }

    companion object {
        @JvmStatic
        protected fun <E> makePages(
            entries: List<E>,
            transformer: EntryTransformer<E>,
            rowPrefixSupplier: RowPrefixSupplier,
            maxEntriesPerPage: Int
        ): Map<Int, MenuPage<E>> {
            val pages: MutableMap<Int, MenuPage<E>> = HashMap()

            var page = 0
            var oldEntry = 0
            val builder = StringBuilder()

            var i = 0
            val entriesSize = entries.size
            while (i < entriesSize) {
                val entry = entries[i]

                val s = transformer.toString(entry)
                Checks.notLonger(s, MessageEmbed.TEXT_MAX_LENGTH - 8, "Entry #$i string")

                if (i - oldEntry >= maxEntriesPerPage || builder.length + s.length > MessageEmbed.TEXT_MAX_LENGTH - 8) {
                    pages[page] = MenuPage(builder.toString(), entries.subList(oldEntry, i))

                    page++
                    oldEntry = i

                    builder.setLength(0)
                }

                builder.append(rowPrefixSupplier.apply(i - oldEntry + 1, maxEntriesPerPage)).append(s).append('\n')
                i++
            }

            pages[page] = MenuPage(builder.toString(), entries.subList(oldEntry, entries.size))

            return pages
        }
    }
}
